### ISSUE TITLE
Can't compose zero-copy `choice(a, b).repeated()` (bug?)

### ISSUE THREAD

**Issue by jerome-trc** [zesterer/chumsky#419]
I recently started using Chumsky trunk so I could help tease out bugs and missing parts of the API, and I came upon something I did a lot in 0.9.2 that doesn't seem to work any more (8ad71861 specifically). The following code only prints ` _`:

```rust
#[test]
fn repro() {
	let parser = primitive::choice((
		primitive::just::<_, _, extra::Default>(" ").map_with_state(|s, _, _| {
			println!("{s}?");
		}),
		primitive::just::<_, _, extra::Default>("_").map_with_state(|s, _, _| {
			println!("{s}!");
		}),
	)).repeated().slice().map_with_state(|s, sp, builder| {
		println!("{s}!");
	});

	let _ = parser.parse(" _");
}
```

I wanted to pass a tree builder as state, but it seems those map calls never happen. I sense this isn't supposed to be the case given the recommendation to "map the output before it goes through `repeated`" given [here](https://github.com/zesterer/chumsky/issues/392#issuecomment-1491623668).

If this is intentional, do I have an alternative? If so I'll submit a PR to write an example or some documentation illustrating how to do what I want without allocating, since I doubt I'm the only one. The closest I could find was in the `map_with_state` part of the doc comment for `foldl_with_state` but it doesn't compile since `map_with_state` doesn't take 2 arguments. Or am I missing something obvious?

Thanks.

**Comment by zesterer** [comment]
Hey. It seems like you're hitting the new GAT-powered optimiser that chumsky uses internally. The behaviour you're observing is correct, even if not immediately intuitive. You can see more details [here](https://github.com/zesterer/chumsky/blob/main/guide/technical_notes.md#purity-and-optimisation) about what's going on.

TL;DR: The `.slice()` call tells chumsky that you don't care about the output value of the parser it's called on, and so chumsky automatically skips anything needed to produce that output value. Although this has surprising results for impure map functions like yours, it's also one of the big things that makes chumsky 1.0 so fast.

**Comment by jerome-trc** [comment]
Oh, good, that's what I was speculating was happening. That means I can work it around immediately. I'll do some cooking; if I end up depending on any really ubiquitous patterns I'll suggest them to see if they belong upstream.

Cheers for the help, and for managing to wrangle the type system into the shape of this crate.
